// @ngxs/store/esm5/src/decorators/select/symbols.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ɵivyEnabled } from '@angular/core';
import { propGetter } from '../../internal/internals';
import { SelectFactory } from './select-factory';
import { throwSelectFactoryNotConnectedError } from '../../configs/messages.config';
/** @type {?} */
var DOLLAR_CHAR_CODE = 36;
/**
 * @template T
 * @param {?} selector
 * @param {?} store
 * @return {?}
 */
export function createSelectObservable(selector, store) {
    // We're doing this stuff to tree-shake the `SelectFactory` when the user
    // is running Ivy since NGXS will select the state from the provided `store` argument.
    return ɵivyEnabled
        ? createSelectObservableIvy(selector, store)
        : createSelectObservableViewEngine(selector);
}
/**
 * @param {?} config
 * @param {?} name
 * @param {?=} rawSelector
 * @param {?=} paths
 * @return {?}
 */
export function createSelectorFn(config, name, rawSelector, paths) {
    if (paths === void 0) { paths = []; }
    rawSelector = rawSelector || removeDollarAtTheEnd(name);
    if (typeof rawSelector !== 'string') {
        return rawSelector;
    }
    else {
        /** @type {?} */
        var propsArray = paths.length
            ? tslib_1.__spread([rawSelector], paths) : rawSelector.split('.');
        return ɵivyEnabled
            ? createSelectorFnIvy(propsArray, config)
            : createSelectorFnViewEngine(propsArray);
    }
}
/**
 * \@example If `foo$` => make it just `foo`
 * @param {?} name
 * @return {?}
 */
export function removeDollarAtTheEnd(name) {
    /** @type {?} */
    var lastCharIndex = name.length - 1;
    /** @type {?} */
    var dollarAtTheEnd = name.charCodeAt(lastCharIndex) === DOLLAR_CHAR_CODE;
    return dollarAtTheEnd ? name.slice(0, lastCharIndex) : name;
}
/**
 * @template T
 * @param {?} selector
 * @param {?} store
 * @return {?}
 */
function createSelectObservableIvy(selector, store) {
    return ngDevMode && !store ? throwSelectFactoryNotConnectedError() : (/** @type {?} */ (store)).select(selector);
}
/**
 * @template T
 * @param {?} selector
 * @return {?}
 */
function createSelectObservableViewEngine(selector) {
    return SelectFactory.store
        ? SelectFactory.store.select(selector)
        : throwSelectFactoryNotConnectedError();
}
/**
 * @param {?} propsArray
 * @param {?} config
 * @return {?}
 */
function createSelectorFnIvy(propsArray, config) {
    return ngDevMode && !config
        ? throwSelectFactoryNotConnectedError()
        : propGetter(propsArray, (/** @type {?} */ (config)));
}
/**
 * @param {?} propsArray
 * @return {?}
 */
function createSelectorFnViewEngine(propsArray) {
    return propGetter(propsArray, (/** @type {?} */ (SelectFactory.config)));
}
